### ISSUE TITLE
3E+ with AtariVox ... not working?

### ISSUE THREAD

**Issue by andrew-davie** [stella-emu/stella#675]
I have this bug that is puzzling me. I took the sample/demo AtariVox code and it runs just fine on Stella+AtariVox.  But when I transplant that code into 3E+ it just does not work. I step through the code, it's doing the same stuff. But the AtariVox doesn't speak.

I'm wondering if this might be a bug in the 3E+ bankswitch code that does not handle output bits to the port properly...?

**Comment by thrust26** [comment]
Does the SaveKey work?

**Comment by andrew-davie** [comment]
> Does the SaveKey work?

I don't know how to test, really.

**Comment by sa666666** [comment]
Still waiting on the 4K ROM you mentioned you would create to test this.

**Comment by andrew-davie** [comment]
The attached binary, though not 4K, is an example of code where the AtariVox works on hardware, but not in Stella.
You should be able to press a joystick direction and hear speech. Works fine via UnoCart on real hardware.
I am absolutely sure I have the AtariVox working in Stella and on hardware.
It is only this 3E+ version that doesn't speak as expected.

[vox3E+.bin.zip](https://github.com/stella-emu/stella/files/4974824/vox3E%2B.bin.zip)

**Comment by andrew-davie** [comment]
The attached **non-3E+** version generally works for me under MacOS, **however** sometimes when I do something as simple as shift simple routines around (and having checked, NO page boundary crossings caused as a result) the speech will stop working.  That is, shifting code around within a page can cause the issue.  And that issue is that the AtariVox goes silent. Only a complete powercycle of my computer will bring it back.  I can go back to the original working code - doesn't work. I unplug/reboot the AtariVox - doesn't work. Reboot the mac -- works. It's almost as if there is some issue with the stella USB code which is leaving it in a non-functional state. Since I'm seeing this on plain vanilla 4K ROMS and 3E+ ROMS I think we can rule out 3E+ code as being the issue.
As noted earlier, **all** of my test versions work just fine on original hardware.

TL;DR... There's possibly something stella-related on MacOS which is intermittent/inconsistent causing USB comms issues to the AtariVox, requiring a machine reboot to fix.

[test.bin.zip](https://github.com/stella-emu/stella/files/4975064/test.bin.zip)

**Comment by thrust26** [comment]
Yes, sounds more like an OS problem now.

@sa666666 Can you verify on MacOS?

**Comment by andrew-davie** [comment]
> Yes, sounds more like an OS problem now.
> 
> @sa666666 Can you verify on MacOS?

Verification would be first running test.bin - it should work.
Then run the vox3E+.bin on stella on MacOS. That one has never worked on stella for me.
Subsequent to that, running test.bin and see how that goes now.

**Comment by sa666666** [comment]
I'll try to test tonight or tomorrow.

**Comment by sa666666** [comment]
OK, I've done some testing.  The 'test.bin' ROM is working fine, as expected.  It is a ROM that we used when adding AVox support to Stella, so it's expected that it would work.

Your 3E+ ROM is also working, but you have to turn on AtariVox for the right controller!  You didn't have to do that for the first test ROM, since Stella already knows about that ROM and has a properties entry for it that enables the AVox.

I went back and forth with the two ROMs, playing sounds from each one.  And stopped and started Stella, and various combinations.  The sound is working in both.

Now, there is one potential difference in our setups.  I don't know if you're using a 2600-daptor flashed to be an AVox interface.  I am not.  I'm using the original "AtariVox USB interface" that was being sold back in the day.  It shouldn't make a difference, but I mention it for completeness.

As for the unplugging/re-plugging, having to reboot the Mac, etc, I am not experiencing that, so I cannot reproduce it.  Is the 3E+ ROM you provided always not working, or does it sometimes work?  You mentioned creating several ROMs after moving banks around.  Changing the timing could be adversely affecting the AVox emulation.  There's code in the emulation that checks if it's been 'too long' since a write, and then forces a new byte to be produced:
```
void AtariVox::clockDataIn(bool value)
{
  if(value && (myShiftCount == 0))
    return;

  // If this is the first write this frame, or if it's been a long time
  // since the last write, start a new data byte.
  uInt64 cycle = mySystem.cycles();
  if((cycle < myLastDataWriteCycle) || (cycle > myLastDataWriteCycle + 1000))
  {
    myShiftRegister = 0;
    myShiftCount = 0;
  }

  // If this is the first write this frame, or if it's been 62 cycles
  // since the last write, shift this bit into the current byte.
  if((cycle < myLastDataWriteCycle) || (cycle >= myLastDataWriteCycle + 62))
  {
    myShiftRegister >>= 1;
    myShiftRegister |= (value << 15);
    if(++myShiftCount == 10)
    {
      myShiftCount = 0;
      myShiftRegister >>= 6;
      if(!(myShiftRegister & (1<<9)))
        cerr << "AtariVox: bad start bit" << endl;
      else if((myShiftRegister & 1))
        cerr << "AtariVox: bad stop bit" << endl;
      else
      {
        uInt8 data = ((myShiftRegister >> 1) & 0xff);
        mySerialPort->writeByte(data);
      }
      myShiftRegister = 0;
    }
  }

  myLastDataWriteCycle = cycle;
}
```
So there are some magic numbers there, and that could definitely be having some effect.  I think the AVox emulation has only been tested on very simple ROMs, and you may be doing (valid) things that are not being correctly emulated.  More work is needed in this area.

Here's the adaptor I'm using:
![IMG_0838](https://user-images.githubusercontent.com/16061066/88485576-601c7600-cf51-11ea-8c57-0f40f5eec2ec.JPG)

**Comment by andrew-davie** [comment]
Thank you for the detailed test. It identified my problem/stupidity and explains everything.
OK, so I was running the standard test program and everything worked fine. Then I'd make a relatively minor change and it didn't work (but did on hardware). Similarly I'd ported the exact same (working) code to 3E+ and it didn't work.
So anyway, to cut a long story short - it was Stella's recognition of the ROM and auto-detecting the profile/atarivox support that threw me. Of course... when I made a change, it was no longer the same ROM.  So although I'd set the AtariVox port in the GUI, I was not passing the port via the command-line call. And thus, Stella was failing because it didn't know the AtariVox was in use/connected.
Now this is one of the first failures I've ever seen of this type - a background thing auto-configuring a parameter which was cruicial for the successful run of a ROM. I'm not saying that's bad - I'm just saying that I missed something cruicial. 
The issues being that although I had set the port via GUI... of course this disappeared when I made changes to the ROM because it was now a different ROM.  Everything now makes sense, and thank you once again for the detailed tests in support of my question/problem.
Now I can stop swearing and get the AtariVox to do it instead. I've just modified Chess and it's speaking as expected now (I'm using the avoxport option on instantiation).

**Comment by sa666666** [comment]
Two things here:
- You don't need to specify the port each time.  You set it once and then forget it.  What needs to be done is tell Stella that an AVox is 'plugged' in as the right controller.  Simply appending `-rc atarivox` should suffice.
- There are still some issues reported with the AVox emulation in Stella.  This is an ongoing thing, so I will close this issue and create another one that is specific to those problems.

**Comment by sa666666** [comment]
Note that this is very similar to developing a paddle game, and Stella not realizing it's using paddles.  So it will default to a joystick and then not work.  It just so happens that the autodetection is smart enough to recognize paddles most of the time, so you don't typically see any problems.  The issue here is that AVox cannot be reliably autodetected, so again it defaults to a joystick.  I ran the ROM from the ROM launcher and saw the problem right away:

![Screenshot_20200727_001414](https://user-images.githubusercontent.com/16061066/88498641-3c335180-cf9e-11ea-80b6-09af1ca63268.png)

**Comment by andrew-davie** [comment]
You have the skills/knowledge/experience that I don't have.
I spent countless hours on this before filing an issue.
Once again, thank you for your support; I really appreciate it.

**Comment by sa666666** [comment]
My point is that you could turn on the ROM launcher too, so you can see the same info :smile: 

Or if you don't use the launcher, have a look in the "System Logs"; the info was there too.  My point isn't that I know how to find the info, it's to explain _how you can find it too_.  To make things easier for you next time.
![Screenshot_20200727_001906](https://user-images.githubusercontent.com/16061066/88498921-0a6eba80-cf9f-11ea-9ddb-5f1c7099f7e1.png)

**Comment by andrew-davie** [comment]
Thank you. I'd never heard of the rom launcher before. Nor was I aware that there was logging available in system logs.
I'll do better next time.

**Comment by thrust26** [comment]
I usually look into the Game Properties for the same reason. :smile: 

@sa666666 How about an option to briefly display the (detected) controllers at startup?

**Comment by sa666666** [comment]
I'm going to close this one, since the problem has been solved.  The AtariVox emulation still needs more work, but that will be addressed in another issue and in a sub-forum on AtariAge.

@thrust26, regarding the display of controllers on startup, please create an issue for that, so we can discuss how it would work.
